// cdk-create-ami/lib/createAMI.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateAMI = exports.ResourceType = exports.VolumeType = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdk = require("aws-cdk-lib");
const constructs_1 = require("constructs");
const customResource_1 = require("./customResource");
var VolumeType;
(function (VolumeType) {
    VolumeType["STANDARD"] = "standard";
    VolumeType["IO1"] = "io1";
    VolumeType["IO2"] = "io2";
    VolumeType["GP2"] = "gp2";
    VolumeType["SC1"] = "sc1";
    VolumeType["ST1"] = "st1";
    VolumeType["GP3"] = "gp3";
})(VolumeType = exports.VolumeType || (exports.VolumeType = {}));
var ResourceType;
(function (ResourceType) {
    ResourceType["IMAGE"] = "image";
    ResourceType["SNAPSHOT"] = "snapshot";
})(ResourceType = exports.ResourceType || (exports.ResourceType = {}));
class CreateAMI extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const uid = cdk.Names.uniqueId(this);
        const { description, instanceId, name, blockDeviceMappings, tagSpecifications, deleteInstance, deleteAmi, } = props;
        const createAMIRequest = new customResource_1.CreateAMICustomResource(this, 'createAMIRequest', {
            uid: uid,
            properties: {
                description: description,
                instanceId: instanceId,
                dryRun: false,
                name: name ?? uid,
                noReboot: false,
                blockDeviceMappings: blockDeviceMappings,
                tagSpecifications: tagSpecifications,
                deleteInstance: deleteInstance ?? false,
                deleteAmi: deleteAmi ?? false,
            },
        });
        this.imageId =
            createAMIRequest.createAMICustomResource.getAttString('imageId');
        this.imageName =
            createAMIRequest.createAMICustomResource.getAttString('imageName');
    }
}
exports.CreateAMI = CreateAMI;
_a = JSII_RTTI_SYMBOL_1;
CreateAMI[_a] = { fqn: "cdk-create-ami.CreateAMI", version: "0.0.116" };
